oak/server/time/roughtime_client.cc

```diff
@@ -0,0 +1,241 @@
+
+// Number of seconds that we will wait for a reply from each server.
+constexpr int kTimeoutSeconds = 3;
+
+// Number of times we will retry connecting to each server.
+constexpr int kServerRetries = 3;
+
+// The size of the receive buffer.
+// This seems to be the same as the minimum request size according to the Roughtime samples.
+// eg. https://roughtime.googlesource.com/roughtime/+/refs/heads/master/simple_client.cc#250
+constexpr size_t kReceiveBufferSize = roughtime::kMinRequestSize;
+
+// Maximum radius accepted for a roughtime response (1 minute in microseconds).
+constexpr uint32_t kMaxRadiusMicroseconds = 60000000;
+
+// Creates a UDP socket connected to the host and port.
+StatusOr<int> CreateSocket(const std::string& host, const std::string& port) {
+  addrinfo hints = {};
+  hints.ai_socktype = SOCK_DGRAM;
+  hints.ai_protocol = IPPROTO_UDP;
+  hints.ai_flags = AI_NUMERICSERV;
+  addrinfo* address;
+  int error = getaddrinfo(host.c_str(), port.c_str(), &hints, &address);
+  if (error != 0) {
+    return Status(asylo::error::GoogleError::INVALID_ARGUMENT,
+                  absl::StrFormat("Could not resolve %s: %s", host.c_str(), gai_strerror(error)));
+  }
+  int handle = socket(address->ai_family, address->ai_socktype, address->ai_protocol);
+  if (handle < 0) {
+    freeaddrinfo(address);
+    return Status(asylo::error::GoogleError::INTERNAL, "Failed to create UDP socket.");
+  }
+  if (connect(handle, address->ai_addr, address->ai_addrlen)) {
+    freeaddrinfo(address);
+    close(handle);
+    return Status(asylo::error::GoogleError::INTERNAL, "Failed to open UDP socket.");
+  }
+  freeaddrinfo(address);
+  return handle;
+}
+
+StatusOr<std::string> SendRequest(const RoughtimeServerSpec& server,
+                                  const Nonce<roughtime::kNonceLength>& nonce) {
+  StatusOr<int> create_socket_result;
+  int attempts = 0;
+  do {
+    ++attempts;
+    create_socket_result = CreateSocket(server.host, server.port);
+  } while (!create_socket_result.ok() && attempts <= kServerRetries);
+
+  if (!create_socket_result.ok()) {
+    return Status(asylo::error::GoogleError::INTERNAL,
+                  "Exceeded maximum retries while attempting to connect to " + server.name);
+  }
+
+  auto handle = create_socket_result.ValueOrDie();
+  const std::string request = roughtime::CreateRequest(nonce.data());
+
+  timeval timeout = {};
+  timeout.tv_sec = kTimeoutSeconds;
+  timeout.tv_usec = 0;
+  setsockopt(handle, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+  ssize_t send_size;
+  do {
+    send_size = send(handle, request.data(), request.size(), 0 /* flags */);
+  } while (send_size == -1 && errno == EINTR);
+  if (send_size < 0 || static_cast<size_t>(send_size) != request.size()) {
+    close(handle);
+    return Status(asylo::error::GoogleError::INTERNAL,
+                  "Network error on sending request to " + server.name);
+  }
+
+  uint8_t receive_buffer[kReceiveBufferSize];
```

Comment:
nit: would making this `const char` rather than `uint8` allow the `reinterpret_cast` below to be removed?